packages/slate/src/interfaces/editor.ts

```diff
@@ -987,100 +986,166 @@ export const Editor = {
       return
     }
 
+    /**
+     * Algorithm notes:
+     * 
+     * Each step `distance` is dynamic depending on the underlying text
+     * and the `unit` specified.  Each step, e.g., a line or word, may
+     * span multiple text nodes, so we iterate through the text both on
+     * two levels in step-sync:
+     * 
+     * `leafText` stores the text on a text leaf level, and is advanced
+     * through using the counters `leafTextOffset` and `leafTextRemaining`.
+     * 
+     * `blockText` stores the text on a block level, and is shortened
+     * by `distance` every time it is advanced.
+     * 
+     * We only keep a window of one blockText and one leafText because
+     * a block node always appears before all of its leaf nodes.
+     */
+
     const range = Editor.range(editor, at)
     const [start, end] = Range.edges(range)
     const first = reverse ? end : start
-    let string = ''
-    let available = 0
-    let offset = 0
+    let blockText = ''
+    let leafTextRemaining = 0
+    let leafTextOffset = 0
+    let leafTextOverflowed = false
     let distance: number | null = null
     let isNewBlock = false
 
+    // Advance blockText and leafText as needed.
     const advance = () => {
+
+      // Advance part 1: Advance blockText, calculate distance.
+      // But only if signalled to do so through setting distance=null.
       if (distance == null) {
         if (unit === 'character') {
-          distance = getCharacterDistance(string)
+          distance = getCharacterDistance(blockText)
         } else if (unit === 'word') {
-          distance = getWordDistance(string)
+          distance = getWordDistance(blockText)
         } else if (unit === 'line' || unit === 'block') {
-          distance = string.length
+          distance = blockText.length
         } else {
           distance = 1
         }
 
-        string = string.slice(distance)
+        blockText = blockText.slice(distance)
       }
 
-      // Add or substract the offset.
-      offset = reverse ? offset - distance : offset + distance
-      // Subtract the distance traveled from the available text.
-      available = available - distance!
-      // If the available had room to spare, reset the distance so that it will
-      // advance again next time. Otherwise, set it to the overflow amount.
-      distance = available >= 0 ? null : 0 - available
+      // Advance part 2: Advance leafText.
+
+      // Step leafText offset and remaining counters by distance.
+      leafTextOffset = reverse ? leafTextOffset - distance : leafTextOffset + distance
+      leafTextRemaining = leafTextRemaining - distance!
```

Comment:
Is the `!` necessary here?

Reply:
It was there before as `distance` was sometimes set to null, but that's no longer the case so I'll remove it.